crates/boundless-market/src/dynamic_gas_filler.rs

```diff
@@ -0,0 +1,146 @@
+    pub fn new(
+        gas_limit_factor: f64,
+        gas_increase_factor: f64,
+        max_gas_multiplier: f64,
+        address: Address,
+    ) -> Self {
+        Self { gas_limit_factor, gas_increase_factor, max_gas_multiplier, address }
+    }
+}
+
+/// Parameters for the dynamic gas filler.
+pub struct DynamicGasParams {
+    /// The fillable gas parameters.
+    pub fillable: GasFillable,
+    /// The multiplier to apply to the gas price.
+    pub multiplier: f64,
+}
+
+impl<N: Network> TxFiller<N> for DynamicGasFiller {
+    type Fillable = DynamicGasParams;
+
+    fn status(&self, tx: &<N as Network>::TransactionRequest) -> FillerControlFlow {
+        TxFiller::<N>::status(&GasFiller, tx)
+    }
+
+    fn fill_sync(&self, _tx: &mut SendableTx<N>) {}
+
+    async fn prepare<P>(
+        &self,
+        provider: &P,
+        tx: &N::TransactionRequest,
+    ) -> TransportResult<Self::Fillable>
+    where
+        P: Provider<N>,
+    {
+        let fillable = GasFiller.prepare(provider, tx).await?;
+
+        let confirmed_nonce = provider.get_transaction_count(self.address).latest().await?;
+        let pending_nonce = provider.get_transaction_count(self.address).pending().await?;
+
+        let tx_diff = pending_nonce.saturating_sub(confirmed_nonce) as u128;
+        tracing::debug!(
+            "DynamicGasFiller: Pending transactions: {}, confirmed transactions: {} - tx_diff: {}",
+            pending_nonce,
+            confirmed_nonce,
+            tx_diff
+        );
+        let mut multiplier = 1.0 + (tx_diff as f64 * self.gas_increase_factor);
+        multiplier = multiplier.min(self.max_gas_multiplier);
+
+        Ok(DynamicGasParams { fillable, multiplier })
+    }
+
+    async fn fill(
+        &self,
+        params: Self::Fillable,
+        mut tx: SendableTx<N>,
+    ) -> TransportResult<SendableTx<N>> {
+        if let Some(builder) = tx.as_mut_builder() {
+            match params.fillable {
+                GasFillable::Legacy { gas_limit, gas_price } => {
+                    let adjusted_gas_limit =
+                        (gas_limit as f64 * (1.0 + self.gas_limit_factor)).ceil() as u64;
+                    let adjusted_gas_price = (gas_price as f64 * params.multiplier) as u128;
+
+                    builder.set_gas_limit(adjusted_gas_limit);
+                    builder.set_gas_price(adjusted_gas_price);
+                    tracing::debug!(
+                        "DynamicGasFiller: Adjusted gas limit: {}, gas price: {}",
+                        adjusted_gas_limit,
+                        adjusted_gas_price
+                    );
+                }
+                GasFillable::Eip1559 { gas_limit, estimate } => {
+                    let adjusted_gas_limit =
+                        (gas_limit as f64 * (1.0 + self.gas_limit_factor)).ceil() as u64;
+                    let adjusted_max_fee =
+                        (estimate.max_fee_per_gas as f64 * params.multiplier) as u128;
```

Comment:
I think we should not adjust the max fee, or adjust it just enough to accommodate the increase to the priority fee. As this is increase here will be quite large compared to the increase to the priority fee.
```suggestion
```